File: tema-3/codigos/Gauss.Seidel.py
import numpy as np

def gauss_seidel(A, b, x0, tol, Nmax):
    """
    Función para resolver un sistema de ecuaciones lineales utilizando el método de Gauss-Seidel.

    Parámetros:
    A (numpy.ndarray): Matriz de coeficientes del sistema.
    b (numpy.ndarray): Vector de términos independientes.
    x0 (numpy.ndarray): Vector de soluciones inicial.
    tol (float): Tolerancia para la convergencia.
    Nmax (int): Número máximo de iteraciones.

    Devuelve:
    numpy.ndarray: Vector de soluciones aproximado.
    int: Número de iteraciones realizadas.
    bool: Indicador de convergencia.
    """
    n = len(b)
    x = x0.copy()
    convergencia = False

    for k in range(Nmax):
        x_ant = x.copy()
        for i in range(n):
            suma = 0
            for j in range(n):
                if j != i:
                    suma += A[i, j] * x[j]
            x[i] = (b[i] - suma) / A[i, i]
        
        # Criterio de convergencia
        if np.linalg.norm(x - x_ant) < tol:
            convergencia = True
            break
    
    return x, k+1, convergencia

# Ejemplo de sistema de ecuaciones lineales
A = np.array([[4, 1, 1],
              [2, 7, 2],
              [1, 2, 4]], dtype=float)
b = np.array([2, 5, 3], dtype=float)
x0 = np.zeros_like(b)
tol = 1e-6
Nmax = 100

# Aplicar el método de Gauss-Seidel
solucion, iteraciones, convergencia = gauss_seidel(A, b, x0, tol, Nmax)

# Imprimir resultados
if convergencia:
    print(f"Solución encontrada: {solucion} en {iteraciones} iteraciones")
else:
    print(f"No se alcanzó la convergencia en {iteraciones} iteraciones")

File: tema-3/codigos/GaussSeidel.java
import java.util.Arrays;

public class GaussSeidel{ 
    /** 
     * @param A     Matriz de coeficientes del sistema.
     * @param b     Vector de términos independientes.
     * @param x0    Vector de soluciones inicial.
     * @param tol   Tolerancia para la convergencia.
     * @param Nmax  Número máximo de iteraciones.
     * @return      Un arreglo que contiene: [Vector de soluciones aproximado, Número de iteraciones, Indicador de convergencia]
     */

     public static Object [] gaussSeidel(double[][] A, double[] b, double[] x0, double tol, int Nmax) {
        int n = b.length; // Número de incógnitas
        double[] x = Arrays.copyOf(x0, n);// Copia del vector inicial para no modificar el original
        boolean convergencia = false;// Indicador de si el método convergió
        int k;//Contador de interaciones
        
        // Iteración principal de Gauss Seidel
        for (k = 0; k < Nmax; k++) {
            double[] x_ant = Arrays.copyOf(x, n); // Guardar la solución anterior para comparar
            
            for (int i = 0; i < n; i++) {
                double suma = 0;
                  // Suma de A[i][j] * x[j], excepto cuando j = i
                for (int j = 0; j < n; j++) {
                    if (j != i) {
                        suma += A[i][j] * x[j];
                    }
                }
                 // Actualización de la solución para la variable i
                x[i] = (b[i] - suma) / A[i][i];
            }
            
            // Verificación del criterio de convergencia: norma de la diferencia < tolerancia
            if (norma(restaVectores(x, x_ant)) < tol) {
                convergencia = true;
                break; // Salir del ciclo si se alcanzó la tolerancia
            }
        }
        
        return new Object[]{x, k + 1, convergencia};
    }
/**
     * Calcula la norma euclidiana de un vector.
     * 
     * @param vector Vector para calcular la norma.
     * @return Norma euclidiana del vector.
     */
    private static double norma(double[] vector) {
        double suma = 0;
        for (double v : vector) {
            suma += v * v;// Suma de cuadrados de los elementos
        }
        return Math.sqrt(suma);// Raíz cuadrada de la suma
    }
    
    /**
     * Resta de dos vectores (a - b).
     * 
     * @param a Primer vector.
     * @param b Segundo vector.
     * @return Vector resultado de la resta.
     */
    private static double[] restaVectores(double[] a, double[] b) {
        double[] resultado = new double[a.length];
        for (int i = 0; i < a.length; i++) {
            resultado[i] = a[i] - b[i];
        }
        return resultado;
    }
     /**
     * Método principal para ejecutar un ejemplo del método de Gauss-Seidel.
     */
    public static void main(String[] args) {
        // Definición del sistema de ecuaciones lineales
        double[][] A = {
            {5, 1, 1},
            {1, 4, 1},
            {1, 1, 6}
        };
        double[] b = {7, 6, 8};
        // Vector inicial x0 (se asume [0, 0, 0])
        double[] x0 = new double[b.length]; // Inicializado con ceros
        double tol = 1e-10; // Tolerancia deseada para la convergencia
        int Nmax = 100;// Máximo número de iteraciones permitidas
        
        // Aplicar el método de Gauss-Seidel
        Object[] resultado = gaussSeidel(A, b, x0, tol, Nmax);
        double[] solucion = (double[]) resultado[0];
        int iteraciones = (int) resultado[1];
        boolean convergencia = (boolean) resultado[2];
        
        // Imprimir resultados
        if (convergencia) {
            System.out.println("Solución encontrada en " + iteraciones + " iteraciones:");
            System.out.println("x = " + String.format("%.6f", solucion[0]));
            System.out.println("y = " + String.format("%.6f", solucion[1]));
            System.out.println("z = " + String.format("%.6f", solucion[2]));
        } else {
            System.out.println("No se alcanzó la convergencia en " + iteraciones + " iteraciones");
        }

    }
}

